zulip_bots/zulip_bots/lib.py

```diff
@@ -102,6 +102,47 @@ def __init__(self, name: str, email: str) -> None:
         self.email = email
         self.mention = '@**' + name + '**'
 
+class CommonStorage(Protocol):
+    def put(self, key: Text, value: Any) -> None:
+        ...
+
+    def get(self, key: Text) -> Any:
+        ...
+
+    def contains(self, key: Text) -> bool:
+        ...
+
+class BotHandler(Protocol):
+
+    email: str
+    full_name: str
+    storage: CommonStorage
+
+    def identity(self) -> BotIdentity:
+        ...
+
+    def react(self, message: Dict[str, Any], emoji_name: str) -> Dict[str, Any]:
+        ...
+
+    def send_message(self, message: Dict[str, Any]) -> Dict[str, Any]:
+        ...
+
+    # ExternalBotHandler and StubBotHandler returns Dict[str, Any]
+    # Whereas EmbeddedBotHandler returns None
+    # Optional[Dict[str, Any]] or object cannot be used because there are occasions where
+    # function calls to send_reply strictly expects Dict[str, Any].
+    def send_reply(self, message: Dict[str, Any], response: str, widget_content: Optional[str] = None) -> Any:
```

Comment:
If there are callers that expect `Dict[str, Any]`, and there are implementation that provide `None`, then that is a bug. The correct resolution isn’t to reach for `Any` and prevent mypy from seeing that there is a bug. The correct resolution is to fix the bug. Either the implementations should be fixed to provide `Dict[str, Any]`, or the callers should be fixed to check for `None`.

Reply:
The `ExternalBotHandler` always returns a `Dict` [see this](https://github.com/zulip/python-zulip-api/blob/3887ad102e26efd1a9e3ed80e95a3ca4ad6600f8/zulip_bots/zulip_bots/lib.py#L161), but the `EmbeddedBotHandler` always returns `None` [see this](https://github.com/zulip/zulip/blob/829f9272d2c4299a0c0a37a09802248d8136c0a8/zerver/lib/bot_lib.py#L111). Our Protocol should be compatible to both I hope. If not then probably I understood it wrong.

Reply:
SO I don't think it is a problem for caller's side